CMake: Multi-Config Generator

ax3l opened the pull request:
Set target properties for multi-config generators.

Fix #1383

Regression to #1326

- [x] test in https://github.com/ax3l/cmake-easyinstall/pull/12

ax3l committed:
CMake: Multi-Config Generator Quickfix

Set target properties for multi-config generators.

ax3l commented:
Hm, does not yet work. Seems to actually collide on a `vc143.pdb` file in the runtime dir...

Update: solved by setting the `COMPILE_PDB_NAME[_CFG]` output name target properties.
Contrary to `PDB_NAME`, the default here is unspecified and seems to default to `vc143.pdb`

franzpoeschel commented on CMakeLists.txt:
```diff
@@ -573,6 +575,19 @@ set_target_properties(openPMD PROPERTIES
+            LIBRARY_OUTPUT_DIRECTORY_${CFG_UPPER} ${openPMD_LIBRARY_OUTPUT_DIRECTORY}
+            RUNTIME_OUTPUT_DIRECTORY_${CFG_UPPER} ${openPMD_RUNTIME_OUTPUT_DIRECTORY}
+            PDB_OUTPUT_DIRECTORY_${CFG_UPPER} ${openPMD_PDB_OUTPUT_DIRECTORY}
+            COMPILE_PDB_OUTPUT_DIRECTORY_${CFG_UPPER} ${openPMD_COMPILE_PDB_OUTPUT_DIRECTORY}
```
```suggestion
            COMPILE_PDB_OUTPUT_DIRECTORY_${CFG_UPPER} ${openPMD_COMPILE_PDB_OUTPUT_DIRECTORY}/${CFG}
```
Weren't the build paths on Windows something like this?

ax3l committed:
CMake: Multi-Config Dir Suffixes

ax3l committed:
Set: `COMPILE_PDB_NAME[_CFG]`

Contrary to `PDB_NAME`, the default here is unspecified
and seems to default to `vc143.pdb`

ax3l commented:
@franzpoeschel ok, now the parallel build works and is not a hack anymore :)

franzpoeschel commented:
Nice!

So, what was missing is the `COMPILE_PDB_NAME` so that the different processes don't try accessing the same file?

ax3l commented:
Yes, exactly. Seems it was unspecified and collided for different executables we build in parallel.
I added an update here: https://github.com/openPMD/openPMD-api/pull/1384#issuecomment-1450968569

franzpoeschel commented:
I actually saw a similar issue in another of my PRs today (the eternal #824), I'll have to rebase and see if this fixes it :)

franzpoeschel reviewed (approved):
Thanks!

ax3l commented on CMakeLists.txt:
```diff
@@ -835,12 +882,27 @@ if(openPMD_HAVE_PYTHON)
+            set_target_properties(openPMD.py PROPERTIES
+                COMPILE_PDB_NAME_${CFG_UPPER} openpmd_api_cxx
+                ARCHIVE_OUTPUT_DIRECTORY_${CFG_UPPER} ${openPMD_PYTHON_OUTPUT_DIRECTORY}/${CFG}/openpmd_api
+                LIBRARY_OUTPUT_DIRECTORY_${CFG_UPPER} ${openPMD_PYTHON_OUTPUT_DIRECTORY}/${CFG}/openpmd_api
```
This broke my slightly arcane logic in `setup.py` for Windows.
Fix in #1400
